### PR TITLE
linux: Improve gpio performance on Raspbery Pi 3

### PR DESCRIPTION
**Pull request by tntclaus**
Improves Raspberry gpio performance up to 70 times through direct memory access.

| Steppers  | Ticks | Rate |
| -----------: | ------------: | ------: |
| 1 stepper | 237 | 211K |
| 2 stepper | 247 | 405K |
| 3 stepper | 257 | 584K |
| 4 stepper | 257 | 778K |
| 5 stepper | 272 | 919K  |
| 6 stepper | 295 | 1017K |

Signed-off-by: Klaus Schwartz <klaus@eraga.net>

### CODE REVIEW

**Review by KevinOConnor — COMMENTED** [review]
Thanks.  See my comments below.

-Kevin

File: src/linux/gpio_bcm.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    if (pull_up > 0) {
+        pud = 2;
+    } else if (pull_up < 0) {
+        pud = 1;
+    }
+
+#if CONFIG_GPIO_NATIVE_RPI3 == 1
+    // Pi 2/3 pull up/down method
+    GPIO_PULL   = pud & 3 ;
+    usleep(5);
```
[the remark this answers is not in the document]
Comment (tntclaus) [reply src/linux/gpio_bcm.c:161]: This code is borrowed from wiringpi, I guess it takes some time to setup PUD for a pin on RPI 2/3. RPI 4 is different. This is tested by me and works both on RPI 3 and 4.

File: src/linux/gpio_bcm.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+#include "autoconf.h"
+#include "gpio.h"
+#include "sched.h" // shutdown
+
+#include <fcntl.h> // open
+#include <stdio.h> // snprintf
+#include <string.h> // memset
+#include <stdlib.h> // atexit
+
+#include "gpio_commons.h"
```
[the remark this answers is not in the document]
Comment (tntclaus) [reply src/linux/gpio_bcm.c:22]: Will fix that

File: src/linux/gpio_bcm.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
+// Includes very efficient support for gpio calls via direct memory
+// access for Raspberry Pi 2/3
+//
+// Signed off to klipper codebase by
+// Klaus Schwartz <klaus@eraga.net> on 19/04/2021.
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
```
[the remark this answers is not in the document]
Comment (tntclaus) [reply src/linux/gpio_bcm.c:7]: Will do that

File: src/linux/Makefile [hunk 4]
```diff
@@ -4,7 +4,7 @@ dirs-y += src/linux src/generic
 
 src-y += linux/main.c linux/timer.c linux/console.c linux/watchdog.c
 src-y += linux/pca9685.c linux/spidev.c linux/analog.c linux/hard_pwm.c
-src-y += linux/i2c.c linux/gpio.c generic/crc16_ccitt.c generic/alloc.c
+src-y += linux/i2c.c linux/$(patsubst "%",%,$(CONFIG_LINUX_GPIO_IMPLEMENTATION)) generic/crc16_ccitt.c generic/alloc.c
```
[the remark this answers is not in the document]
Comment (tntclaus) [reply src/linux/Makefile:7]: Got it

File: src/linux/Kconfig [hunk 5]
```diff
@@ -21,4 +21,26 @@ config CLOCK_FREQ
...
+config LINUX_GPIO_IMPLEMENTATION
+    string
+    default "gpio_generic.c"
+    depends on GPIO_GENERIC
+
+config LINUX_GPIO_IMPLEMENTATION
+    string
+    default "gpio_bcm.c"
+    depends on GPIO_NATIVE_RPI3 || GPIO_NATIVE_RPI4
+
```
[the remark this answers is not in the document]
Comment (tntclaus) [reply src/linux/Kconfig:45]: Understood. Will prepare fixes in few days

**Comment by tntclaus** [comment]
While using for several days my printer I noticed that stepper to far in past shutdown may still get triggered when:

* RPI 3 CPU temp gets above 60°C;
* File is uploaded to SD card via Ethernet.

This doesn't depend on print speed.

At the same time this happened only 3 times with me. I hope IRQ branch will bring improvements to this.

**Comment by tntclaus** [comment]
Kevin, this is for review only, do not merge yet. Seems like RPi requires RT kernel to maintain 1000us timing window (which is what I am testing as of today), `isolcpus` kernel setting process scheduler to RR, priority to highest and cpu affinity to isolated core, otherwise random tasks may interrupt mcu at any time delaying timer for more than 1ms. 

Probably it will take about a week of successful printing without "stepper too far in past" error to confirm the patch works as intended.

**Comment by KevinOConnor** [comment]
Thanks.  A few comments:
1. I find the header files to be confusing.  I think it would be preferable to add all the code to a single new bcm27xx_gpio.c file.
2. If there is a change needed to the klippy_start.sh script I think it should be in a separate PR.
3. It would be preferable to avoid #ifdefs where possible.  For example, the gpio_in_reset() code can use `if (CONFIG_MACH_LINUX_BCM2709)` to avoid an ifdef.  The advantage of this coding style is that we get better compiler error checking.
4. The commented out definitions in Kconfig should be removed.
5. I think we can merge the gpio improvements separately from the scheduling concerns.  Having faster gpio access shouldn't harm anything.

Cheers,
-Kevin

**Comment by KevinOConnor** [comment]
As an additional comment, I'm not sure I would bother using the `ioctl()` interface if direct memory access is being used.  It makes the code more complicated and I'm not sure it provides a benefit.

Cheers,
-Kevin

**Comment by tntclaus** [comment]
Okay, will do that in next few days.

**Comment by KevinOConnor** [comment]
Any further updates on this?

-Kevin

**Comment by KevinOConnor** [comment]
It looks like this ticket has become stale, so I'm closing it for now.

-Kevin

**Pull request closed without merging**